beacon-chain/p2p/crawler.go

```diff
@@ -0,0 +1,387 @@
+		case <-c.ctx.Done():
+			return
+		}
+	}
+}
+
+// crawl performs a single crawl of the discovery network.
+func (c *Crawler) crawl() {
+	ctx, cancel := context.WithTimeout(c.ctx, c.crawlTimeout)
+	defer cancel()
+
+	iterator := c.dv5.RandomNodes()
+	// Ensure iterator unblocks on context cancellation or timeout
+	go func() {
+		<-ctx.Done()
+		iterator.Close()
+	}()
+
+	for iterator.Next() {
+		if ctx.Err() != nil {
+			return
+		}
+		node := iterator.Node()
+		if node == nil {
+			continue
+		}
+
+		if !c.service.filterPeer(node) {
+			c.RemovePeer(node.ID())
+			continue
+		}
+
+		topics, err := c.topicExtractor(ctx, node)
+		if err != nil {
+			// It's okay for the callback to error, just means we can't process this node.
+			log.WithError(err).WithField("node", node.ID()).Debug("Failed to extract topics from node")
+			continue
+		}
+		if len(topics) == 0 {
+			// If no topics are returned, we don't track the peer.
+			// We should also remove it if it's already tracked.
+			c.RemovePeer(node.ID())
+			continue
+		}
+		c.addOrUpdatePeer(node, topics)
+	}
+}
+
+// PeersForTopic returns a slice of peers known to be associated with the given topic.
+func (c *Crawler) PeersForTopic(topic string) []*enode.Node {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+
+	peers, ok := c.topicToPeers[topic]
+	if !ok {
+		return nil
+	}
+	var result []*enode.Node
+	for enodeID := range peers {
+		if pNode, ok := c.peerNodes[enodeID]; ok {
+			result = append(result, pNode.node)
+		}
+	}
+	return result
+}
+
+// RemoveTopic removes all peer associations for a given topic. This is useful when
+// a beacon node unsubscribes from a topic and no longer needs to track its peers.
+func (c *Crawler) RemoveTopic(topic string) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	peers, ok := c.topicToPeers[topic]
+	if !ok {
+		return
+	}
+	delete(c.topicToPeers, topic)
+
+	for enodeID := range peers {
+		pNode, ok := c.peerNodes[enodeID]
+		if !ok {
+			continue
+		}
+		// Remove the topic from the peer's topic list.
+		var newTopics []string
+		for _, t := range pNode.topics {
+			if t != topic {
+				newTopics = append(newTopics, t)
+			}
+		}
+		pNode.topics = newTopics
+
+		// If the peer is no longer in any topics, remove it entirely.
+		if len(pNode.topics) == 0 {
+			c.removePeerUnlocked(enodeID)
+		}
+	}
+}
+
+func (c *Crawler) RemovePeer(enodeID enode.ID) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	c.removePeerUnlocked(enodeID)
+}
+
+func (c *Crawler) removePeerUnlocked(enodeID enode.ID) {
+	pNode, ok := c.peerNodes[enodeID]
+	if !ok {
+		return
+	}
+	delete(c.peerIDToEnodeID, pNode.peerID)
+	delete(c.peerNodes, enodeID)
+
+	for _, topic := range pNode.topics {
+		if peers, ok := c.topicToPeers[topic]; ok {
+			delete(peers, enodeID)
+			if len(peers) == 0 {
+				delete(c.topicToPeers, topic)
+			}
+		}
+	}
+}
+
+// RemovePeerByPeerID removes a peer using its libp2p peer.ID.
+func (c *Crawler) RemovePeerByPeerID(pid peer.ID) {
```

Comment:
nitpick: does the naming track for you if this was called `RemovePeerId`? I feel like it would be less verbose and still intuitive to understand that we're removing all data related to that peer (since we're not just storing peer ids for their own sake).